File: assets/examples/2022/scala3-aws-lambda-dynamodb-importer/src/main/scala/Handler.scala
import com.amazonaws.services.lambda.runtime.events.{APIGatewayV2HTTPEvent, APIGatewayV2HTTPResponse}
import com.amazonaws.services.lambda.runtime.{Context, LambdaLogger, RequestHandler}
import software.amazon.awssdk.auth.credentials.{AwsCredentialsProvider, DefaultCredentialsProvider, ProfileCredentialsProvider}
import software.amazon.awssdk.core.SdkBytes
import software.amazon.awssdk.profiles.ProfileFile
import software.amazon.awssdk.protocols.jsoncore.{JsonNode, JsonNodeParser}
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.dynamodb.DynamoDbClient
import software.amazon.awssdk.services.dynamodb.model.*

import java.nio.file.Path
import java.util.Base64
import scala.jdk.CollectionConverters.*
import scala.jdk.OptionConverters.*
import scala.util.{Failure, Success, Try}

val tableName = "demo_stock_prices"

val dynamoDbClient = DynamoDbClient.builder
  .credentialsProvider(DefaultCredentialsProvider.create)
  .region(Region.US_EAST_1)
  .build

class Handler extends RequestHandler[APIGatewayV2HTTPEvent, APIGatewayV2HTTPResponse] :

  override def handleRequest(event: APIGatewayV2HTTPEvent, context: Context): APIGatewayV2HTTPResponse =
    given lambdaLogger: LambdaLogger = context.getLogger

    lambdaLogger.log("Start")
    lambdaLogger.log(event.getBody)

    val body = Option(event.getBody).withFilter(!_.isBlank)
      .map(Success(_))
      .getOrElse(Failure(ParseException("empty body", "''")))

    body
      .flatMap(parseStockPriceItems)
      .flatMap(putIntoDynamoDB)
      .fold(errorToResult, count =>
        APIGatewayV2HTTPResponse.builder
          .withStatusCode(200)
          .withBody(s"""{ "added": $count }""")
          .build
      )

case class ParseException(error: String, content: String) extends Exception(error)

def parseStockPriceItems(json: String)(using lambdaLogger: LambdaLogger): Try[Iterable[StockPriceItem]] =
  lambdaLogger.log(json)
  Try(JsonNodeParser.create.parse(json).asArray.asScala.map(StockPriceItem.apply))

def putIntoDynamoDB(stockPriceItems: Iterable[StockPriceItem])(using lambdaLogger: LambdaLogger): Try[Long] = Try {
  val writeRequests = stockPriceItems.map {
    stockPriceItem =>
      val request = PutRequest.builder.item(stockPriceItem.dynamoDBAttributeMap).build
      WriteRequest.builder.putRequest(request).build
  }
  val requestItems = Map(tableName -> writeRequests.toList.asJava).asJava
  val batchWriteItemRequest = BatchWriteItemRequest.builder.requestItems(requestItems).build
  val batchWriteItemResponse = dynamoDbClient.batchWriteItem(batchWriteItemRequest)
  if (batchWriteItemResponse.hasUnprocessedItems && batchWriteItemResponse.unprocessedItems.size > 0) {
    val message = s"Wrote ${writeRequests.size - batchWriteItemResponse.unprocessedItems.size} of ${writeRequests.size}"
    throw new Exception(message)
  } else {
    lambdaLogger.log("Success")
    writeRequests.size
  }
}

def errorToResult(ex: Throwable)(using lambdaLogger: LambdaLogger): APIGatewayV2HTTPResponse =
  ex match
    case ParseException(error, content) =>
      val message = s"Error parsing request $error in $content"
      lambdaLogger.log(message)
      APIGatewayV2HTTPResponse.builder.withStatusCode(400).withBody(message).build
    case _ =>
      throw ex

File: examples/2022/scala3-aws-lambda-dynamodb-importer/src/main/python/handler.py
import boto3
import json

import json

TABLE_NAME = 'demo_stock_prices'

dynamodb_table = boto3.resource('dynamodb').Table(TABLE_NAME)

def lambda_handler(event, context):
    try:
        if 'body' not in event:
            raise ParseException('empty body', "''")
        body = event['body']
        stock_price_items = parse_stock_price_items(body)
        result = put_into_dynamodb(stock_price_items)
        return {
            'statusCode': 200,
            'body': json.dumps({ "added": result })
        }
    except ParseException as ex:
        return error_to_result(ex)

class ParseException(RuntimeError):
    def __init__(self, error, content):
        self.error = error
        self.content = content

def parse_stock_price_items(str):
    print(str)
    stock_price_items = json.loads(str)
    for item in stock_price_items:
        if 'symbol' not in item:
            raise ParseException('`symbol` not found', item)
        if 'time' not in item:
            raise ParseException('`time` not found', item)
        if 'prices' not in item:
            raise ParseException('`prices` not found', item)
    return stock_price_items


def put_into_dynamodb(stock_price_items):
    try:
        with dynamodb_table.batch_writer() as batch:
            for item in stock_price_items:
                batch.put_item(Item=item)

        print('Success')
        return len(stock_price_items)
    except:
        message = f"Wrote 0 of {len(stock_price_items)}"
        print(message)
        raise

def error_to_result(ex):
    message = f"Error parsing request {ex.error} in {ex.content}"
    print(message)
    return {
        'statusCode': 400,
        'body': message
    }